kms/src/main/java/io/kroxylicious/kms/service/Kms.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.service;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.crypto.SecretKey;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Abstracts the KMS operations needed for Envelope Encryption
+ * @param <K> The type of Key Encryption Key id.
+ * @param <E> The type of encrypted Data Encryption Key.
+ */
+public interface Kms<K, E> {
```

Comment:
Is the name of this really correct? We we envisage it evolving all the operations, or should be remain specific to envelope encryption, with other interfaces obtained from `KmsService` for other operations (such as key provisioning)? That latter model would be a bit more flexible. Not all KMSs might support all operations, for example. Or you could pass a `EnvelopeEncryptionService` to a method and know it's not going to be able to create/delete keys.

Reply:
I agree. Needing other key operations, such as `createKey` is an easy to imagine future.  And yes, some providers might not be able to support all classes of operations, so I like the decompose into separate interfaces idea a lot.

However, I think what we have here is good enough, so I'd stick with it for at least 0.4.0.